### DISCUSSION TITLE
Updating UI with progress during long method calls

### DISCUSSION THREAD

**Discussion by dnnp2011** [livewire/livewire#5114] [1. Help] [answered]
I've seen other people ask this, but haven't found a satisfactory answer:
My Livewire component triggers a long-running process that needs to go through multiple steps, and I want to push messages to a user-facing console when progress is made. Currently, however, the console messages are not rendered until the entire process completes. Which obviously defeats the purpose.

I've tried using Livewire and browser events in various ways, but they have the same limitation - Livewire doesn't emit them until the method finishes running. I've tried using various Livewire directives in a plethora of ways, to no avail.

The only answer I've seen that comes close to touching this problem is the recommendation to put the long-running process in a Job and just dispatch the job from my Livewire component. And this might solve the problem of a long-running Livewire method, but introduces a handful of other problems, like forcing me to persist the API request to a database or cache (I want most of the response to be ephemeral), forcing me to have to resort to polling for when the data is ready or implementing a websocket, forcing me to either chain together multiple jobs for different steps of the process or put all the steps in a single job and face the same problem of progress being opaque.

Is there really not a better way? Was I mistaken in thinking Livewire was a reactive, dynamic UI framework? Isn't this a pretty quintessential reactive UI scenario?

**Comment by joshhanley** [comment, marked as answer]
@dnnp2011 due to the nature of PHP and the fact that a Livewire request just makes a request to the server, which returns a single response, Livewire can't display progress updates without the request being finalised. The reason is, it requires the server to render the complete html and return it to the browser, so the page can be updated with the changes.

The best way would be to use a job and polling as you suggested. Or you can use websockets as per this article by @PhiloNL https://philo.dev/laravel-batches-and-real-time-progress-with-livewire/

Livewire is more dynamic than a static html page, and it can enhance elements on the page and make them reactive, but not to the point of handling a long running process and things like that. Livewire isn't a reactive/dynamic UI framework in that sense, for that you'd be better off using a front end framework.

Hope this helps!

**Reply by dnnp2011** [reply]
@joshhanley, thanks for enlightening me about the limitations and constraints of Livewire.
I don't like just being told "it can't be done", but I can accept a reasonable explanation from a knowledgeable person.

I really do like Livewire, I was just experiencing a moment of frustration trying to solve this particular problem.
Thanks for your hard work 😉

**Reply by joshhanley** [reply]
@dnnp2011 no worries! 🙂 yeah fair enough, glad I could help! Thanks for sharing your solution too 😁

**Comment by dnnp2011** [comment]
For posterity, the way I ended up working around this limitation, is by breaking up the steps into their own Livewire methods, then calling each step in turn from within a javascript function defined in Alpine's `x-data` directive. This allowed me to display a new log message before and after each step, without dealing with jobs, websockets, polling, and all that stuff.

For example:

```html
<div
    x-data="{
    async runProcess() {
        $wire.set('isRunning', true);
    
        // Step 1: Search API
        await this.pushToLog('Searching API...', this.LEVEL.NOTICE);
        const resultCount = await $wire.call('runApiSearch');
        await this.pushToLog('Done searching API', this.LEVEL.NOTICE);
    
        // Step 2: Format results into required shape
        await this.pushToLog('Processing results...', this.LEVEL.NOTICE);
        const formattedResults = await $wire.call('runFormatResults');
        if (!formattedResults) {
            return $wire.set('isRunning', false);
        }
        await this.pushToLog('Done processing results', this.LEVEL.NOTICE);
    
        // Step 3: Resolve minified URLs
        await this.pushToLog('Resolving minified URLs...', this.LEVEL.NOTICE);
        const resolvedUrls = await $wire.call('runResolveUrls');
        if (!resolvedUrls) {
            return $wire.set('isRunning', false);
        }
        await this.pushToLog('Done resolving URLs', this.LEVEL.NOTICE);
    
        $wire.set('isRunning', false);
    },
   } "
>

...

    <button @click="runProcess">Start Long Running Process</button>
</div>
```
